Applications/Spire/Source/Ui/PopupBox.cpp

```diff
@@ -0,0 +1,182 @@
+#include "Spire/Ui/PopupBox.hpp"
+#include <QApplication>
+#include "Spire/Ui/Layouts.hpp"
+
+using namespace boost::signals2;
+using namespace Spire;
+
+PopupBox::PopupBox(QWidget& body, QWidget* parent)
+    : QWidget(parent),
+      m_body(&body),
+      m_window(nullptr),
+      m_focus_observer(*this),
+      m_body_focus_observer(*m_body),
+      m_position_observer(*this),
+      m_alignment(Alignment::NONE),
+      m_min_height(m_body->sizeHint().height()),
+      m_max_height(QWIDGETSIZE_MAX),
+      m_above_space(0),
+      m_below_space(0),
+      m_right_space(0),
+      m_focus_connection(m_focus_observer.connect_state_signal(
+        std::bind_front(&PopupBox::on_focus, this))),
+      m_body_focus_connection(m_body_focus_observer.connect_state_signal(
+        std::bind_front(&PopupBox::on_body_focus, this))),
+      m_position_connection(m_position_observer.connect_position_signal(
+        std::bind_front(&PopupBox::on_position, this))) {
+  m_body->installEventFilter(this);
+  m_body->setSizePolicy(QSizePolicy::Expanding, QSizePolicy::Expanding);
+  enclose(*this, *m_body);
+}
+
+const QWidget& PopupBox::get_body() const {
+  return *m_body;
+}
+
+QWidget& PopupBox::get_body() {
+  return *m_body;
+}
+
+QSize PopupBox::sizeHint() const {
+  if(has_popped_up()) {
```

Comment:
Can you describe the goal of this method? It looks like it could be rewritten as:

    if(has_popped_up()) {
      return m_body->sizeHint();
    }
    return QWidget::sizeHint();

Reply:
The goal of the method is to keep the PopupBox the same size as before the body popping up. When the body poppes out from the layout of the PopupBox, the size of PopupBox will become zero, so the method prevents it from becoming zero. I don’t think the rewritten code is same as it. 
 m_body->sizeHint() can return different sizes as the content changes, while the method needs to return a constant size which is the size before the body popping up. This constant size is stored in m_size_hint.